dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/sms/listener/CompressionSMSListener.java

```diff
@@ -0,0 +1,329 @@
+        this.trackedEntityTypeService = trackedEntityTypeService;
+        this.trackedEntityAttributeService = trackedEntityAttributeService;
+        this.programService = programService;
+        this.organisationUnitService = organisationUnitService;
+        this.categoryService = categoryService;
+        this.dataElementService = dataElementService;
+        this.programStageInstanceService = programStageInstanceService;
+        this.identifiableObjectManager = identifiableObjectManager;
+    }
+
+    @Override
+    public boolean accept( IncomingSms sms )
+    {
+        if ( sms == null || !SmsUtils.isBase64( sms ) )
+        {
+            return false;
+        }
+
+        SMSSubmissionHeader header = getHeader( sms );
+        if ( header == null )
+        {
+            // If the header is null we simply accept any listener
+            // and handle the error in receive() below
+            return true;
+        }
+        return handlesType( header.getType() );
+    }
+
+    @Override
+    public void receive( IncomingSms sms )
+    {
+        SMSSubmissionReader reader = new SMSSubmissionReader();
+        SMSSubmissionHeader header = getHeader( sms );
+        if ( header == null )
+        {
+            // Error with the header, we have no message ID, use -1
+            sendSMSResponse( SMSResponse.HEADER_ERROR, sms, -1 );
+            return;
+        }
+
+        SMSMetadata meta = getMetadata( header.getLastSyncDate() );
+        SMSSubmission subm = null;
+        try
+        {
+            subm = reader.readSubmission( SmsUtils.getBytes( sms ), meta );
+        }
+        catch ( Exception e )
+        {
+            log.error( e.getMessage() );
+            sendSMSResponse( SMSResponse.READ_ERROR, sms, header.getSubmissionID() );
+            return;
+        }
+
+        // TODO: Can be removed - debugging line to check SMS submissions
+        Gson gson = new GsonBuilder().setPrettyPrinting().create();
+        log.info( "New received SMS submission decoded as: " + gson.toJson( subm ) );
+
+        SMSResponse resp = null;
+        try
+        {
+            checkUser( subm );
+            resp = postProcess( sms, subm );
+        }
+        catch ( SMSProcessingException e )
+        {
+            log.error( e.getMessage() );
+            sendSMSResponse( e.getResp(), sms, header.getSubmissionID() );
+            return;
+        }
+
+        log.info( "SMS Response: " + resp.toString() );
+        sendSMSResponse( resp, sms, header.getSubmissionID() );
+    }
+
+    private void checkUser( SMSSubmission subm )
+    {
+        UID userid = subm.getUserID();
+        User user = userService.getUser( userid.uid );
+
+        if ( user == null )
+        {
+            throw new SMSProcessingException( SMSResponse.INVALID_USER.set( userid ) );
+        }
+    }
+
+    private SMSSubmissionHeader getHeader( IncomingSms sms )
+    {
+        byte[] smsBytes = SmsUtils.getBytes( sms );
```

Comment:
Just a suggestion. If SmsUtils is unable to decode text then it will raise an exception which would result in storing NULL in smsBytes.

Reply:
See [here](https://github.com/dhis2/dhis2-core/pull/3600/files#diff-73a13d57ea7ccac3231da4abff3a4e19R143)

There is a check to confirm that the text is Base64, if it isn't this listener will skip it. If it is, we can be fairly sure getBytes will be fine.